src/serializer/serializer.js

```diff
@@ -1333,97 +1360,161 @@ export class Serializer {
           define(instance, funcNode);
         }
       } else {
-        let suffix = instances[0].functionValue.__originalName || "";
-        let factoryId = t.identifier(this.factoryNameGenerator.generate(suffix));
-
-        // filter included variables to only include those that are different
-        let factoryNames: Array<string> = [];
-        let sameSerializedBindings = Object.create(null);
-        for (let name in names) {
-          let isDifferent = false;
-          let lastBinding;
-
-          for (let { serializedBindings } of instances) {
-            let serializedBinding = serializedBindings[name];
-            invariant(!serializedBinding.modified);
-            if (!lastBinding) {
-              lastBinding = serializedBinding;
-            } else if (!AreSameSerializedBindings(serializedBinding, lastBinding)) {
-              isDifferent = true;
-              break;
+        // Group instances with modified bindings
+        let instanceBatches = [instances];
+        for (let name in modified) {
+          let newInstanceBatches = [];
+
+          for (let batch of instanceBatches) {
+            // Map from representative binding to function instances that use it
+            let bindingLookup = new Map();
+
+            for (let functionInstance of batch) {
+              let serializedBinding = functionInstance.serializedBindings[name];
+              let found = false;
+              for (let [binding, group] of bindingLookup.entries()) {
+                if ((serializedBinding.value &&
+                  binding.value && SameValuePartial(this.realm, serializedBinding.value, binding.value)) ||
```

Comment:
`SameValuePartial` can throw on abstract values. Don't use that. Instead, check if the values are `instanceof ConcreteValue`, and then call `SameValue`.
I think it would be best if you "upgraded" the implementation of `AreSameSerializedBindings` to do the `SameValue` check on the `.value` components, instead of spelling that all out here (and not in the other usage of `AreSameSerializedBindings`).